### PR TITLE
mgr/dashboard: Improve level A accessibility for pagination component

### PR DESCRIPTION
**Pull request by nSedrickm**
Add custom pagination component based on patternfly design

Before
image

After
image

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by epuertat — APPROVED** [review]
Awesome, @nSedrickm ! Everything looks good to me, but I'd first like to ensure that the Dashboard community agrees with bringing Patternfly in the Dashboard, since we already have Bootstrap, and Patternfly is a bit heavyweight.

@nizamial09 what do you think?

@s0nea would this have any impact to your downstream branding? In our case we alrady used Patternfly for downstream, so this would be seamless.

If we introduce Patternfly, probable we should discuss about the future of Bootstrap in the Dashboard arena, since IMHO Patternfly better provides UI styling and widgets for a website like the Dashboard...

**Review by nizamial09 — COMMENTED** [review]
@nSedrickm I like the fact that patternfly is getting applied to upstream. But TBH I don't like it being applied for just one component and leaving it at that.

If we are introducing patternfly here its going to break some things in the dashboard like the search bar in table, and some minor UI elements, some parts of the form. Also this makes somethings in the dashboard being bigger while somethings being smaller and that causes inconsistency between the UI elements. It looked to me like a minorly broken dashboard. 

And I was checking the size of the bundle generated and it was around 43M with patternfly there. But the interesting this is it was 40M without the patternfly. That should not be ideal.

And we shouldn't keep bootstrap and patternfly together. I think if we go for patternfly, then we should gradually remove the bootstrap. we can probably discuss this in fridays standup.

@bryanmontalvan @Pegonzal might be interesting for you guys too!
Some broken elements.
Screenshot from 2022-08-17 20-55-20
Screenshot from 2022-08-17 20-54-41
table
scrnli_17_08_2022_20-57-41
scrnli_17_08_2022_20-56-11

File: src/pybind/mgr/dashboard/frontend/angular.json [hunk 1]
```diff
@@ -56,6 +56,7 @@
               }
             ],
             "styles": [
+              "node_modules/@patternfly/patternfly/patternfly.min.css",
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/angular.json:59]: i am thrilled, but there are concerns though

**Comment by nSedrickm** [comment]
Thanks @nizamial09, @epuertat. I initially wanted to only include the styling needed for the pagination component like in the g[etting started docs](https://www.patternfly.org/v4/get-started/develop#htmlcss) but it didn't work so I resulted to importing the whole css file. Perhaps I am missing something

**Comment by epuertat** [comment]
@nSedrickm personally I'd love to see Patternly (or any other UI that improves the Bootstrap styling) in place, but as Nizam pointed out, it conflicts with the current Boostratp UI causing those quirky appearances.

What's important from the Patternfly pagination is not the UI itself, but the UX that it provides: users can access any page by typing the page number:

image

With the current pagination it might require hundreds of clicks to navigate a table with thousands of elements (and hence hundreds of pages):

image

The other Patternfly widget is not really required (yet):

image

Could you please try to achieve a similar (it doesn't have to be exactly the same) look as the Patternfly one?

**Comment by nSedrickm** [comment]
> @nSedrickm personally I'd love to see Patternly (or any other UI that improves the Bootstrap styling) in place, but as Nizam pointed out, it conflicts with the current Boostratp UI causing those quirky appearances.
> 
> What's important from the Patternfly pagination is not the UI itself, but the UX that it provides: users can access any page by typing the page number:
> 
> image
> 
> With the current pagination it might require hundreds of clicks to navigate a table with thousands of elements (and hence hundreds of pages):
> 
> image
> 
> The other Patternfly widget is not really required (yet):
> 
> image
> 
> Could you please try to achieve a similar (it doesn't have to be exactly the same) look as the Patternfly one?

I now understand. Sure, I can replicate the design from patternfly with the current tooling

**Review by epuertat — CHANGES_REQUESTED** [review]
Hey Sedrick, great work here!

I see some formatting issues (the left-most text font size is 12 px, while the right-most is 13.2 px.
image

Removing all the custom styling it doesn't look that bad (and it's consistent with the current styling), right?
image

On the RBD page (the one with server-side validation), I think that this change breaks that part (you can easily reproduce it by creating 4 RBD images and setting the number of items/page to 2):

image

That picture should show at least 2 pages. @pereman2 I think that @nSedrickm might need your help here!

**Comment by nSedrickm** [comment]
Hi @epuertat, made some updates here. The code for controlling the page limit has not been changed. Wondering if we still need to add debounce

**Review by epuertat — APPROVED** [review]
LGTM! Great job @nSedrickm ! Thank you for addressing my comments.

image

**Review by nizamial09 — APPROVED** [review]
Awesome job with the new pagination component! Thanks @nSedrickm 

Just one [unit test failure](https://jenkins.ceph.com/job/ceph-pull-requests/101982/consoleFull#-17253971136733401c-e9d0-4737-9832-6594c5da0afa)

```
● RgwDaemonListComponent › should show a row with daemon info

    expect(received).toContain(expected) // indexOf

    Expected substring: "total 1"
    Received string:    "IDHostnameZoneZone GroupRealmVersionIDHostnameZoneZone GroupRealmVersion 0 total of 1 "

      77 |     expect(listDaemonsSpy).toHaveBeenCalledTimes(1);
      78 |     expect(component.daemons).toEqual([daemon]);
    > 79 |     expect(fixture.debugElement.query(By.css('cd-table')).nativeElement.textContent).toContain(
         |                                                                                      ^
      80 |       'total 1'
      81 |     );
      82 |

      at src/app/ceph/rgw/rgw-daemon-list/rgw-daemon-list.component.spec.ts:79:86
      at node_modules/zone.js/dist/fake-async-test.js:610:34
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:386:30)
      at ProxyZoneSpec.Object.<anonymous>.ProxyZoneSpec.onInvoke (node_modules/zone.js/dist/proxy.js:117:43)
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:385:36)
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/dist/zone.js:143:47)
```

**Comment by ljflores** [comment]
@nSedrickm this is where the test is: https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-daemon-list/rgw-daemon-list.component.spec.ts#L74-L81

**Comment by nSedrickm** [comment]
Thanks @nizamial09, @ljflores the reason the test is failing is a difference in "of" between total and 1 🤣

```
Expected substring: "total 1"
Received string:    "IDHostnameZoneZone GroupRealmVersionIDHostnameZoneZone GroupRealmVersion 0 total of 1 "
```

**Comment by nSedrickm** [comment]
@epuertat did some testing here. Currently the debounce interval for server [side pagination changes](https://github.com/ceph/ceph/blob/f5c21acfdc16e8a519f58dccee808baf0c713c4d/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts#L269-L275) is 1s. At lower values the delay is not really effective  and above one second too noticeable.  According to these aricles [one](https://www.nngroup.com/articles/response-times-3-important-limits/#:~:text=The%20basic%20advice%20regarding%20response,except%20to%20display%20the%20result.), [two](https://medium.com/@slhenty/ui-response-times-acec744f3157) 1s seems to be a good value.

**Comment by nizamial09** [comment]
jenkins test api

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table-pagination/table-pagination.component.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    }
+  }
+
+  calcPages(page?: number): any[] {
+    const pages = [];
+    let startPage = 1;
+    let endPage = this.totalPages;
+    const maxSize = 5;
+    const isMaxSized = maxSize < this.totalPages;
+
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table-pagination/table-pagination.component.ts:85]: This is a nitpick from my side but you don't need a list of pages to know in which page you are. `offset = page_num * limit_per_page` where `page_num >= 0 && page_num < max_pages`

**Comment by epuertat** [comment]
jenkins test api

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table-pagination/table-pagination.component.ts (hunk 2 above)
Comment (nSedrickm) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table-pagination/table-pagination.component.ts:85]: Thanks @pereman2, Is this the offset from the main table component ?

**Comment by pereman2** [comment]
jenkins test api

**Pull request merged**